common/lib/capa/capa/responsetypes.py

```diff
@@ -831,16 +841,110 @@ def assign_choice_names(self):
+        elif halves_error_count == 1 and len(all_choices) > 2:
+            return_grade = round(self.get_max_score() / 2.0, 2)
+            return CorrectMap(self.answer_id, correctness='partially-correct', npoints=return_grade)
+        elif halves_error_count == 2 and len(all_choices) > 4:
+            return_grade = round(self.get_max_score() / 4.0, 2)
+            return CorrectMap(self.answer_id, correctness='partially-correct', npoints=return_grade)
+        else:
+            return CorrectMap(self.answer_id, 'incorrect')
+
     def get_score(self, student_answers):
 
+        # Setting up answer lists:
+        #  all_choices: the full list of checkboxes
+        #  student_answer: what the student actually chose (note no "s")
+        #  student_non_answer: what they didn't choose
+        #  self.correct_choices: boxes that should be checked
+        #  self.incorrect_choices: boxes that should NOT be checked
+
+        all_choices = self.correct_choices.union(self.incorrect_choices)
+
         student_answer = student_answers.get(self.answer_id, [])
 
         if not isinstance(student_answer, list):
             student_answer = [student_answer]
 
+        # "None apply" should really be a valid choice for "check all that apply",
+        # but it throws an error if all the checks are blank.
         no_empty_answer = student_answer != []
+
         student_answer = set(student_answer)
 
+        student_non_answers = all_choices - student_answer
+
+        if not no_empty_answer:
+            return CorrectMap(self.answer_id, 'incorrect')
+
+        # This below checks to see whether we're using an alternate grading scheme.
+        #  Set partial_credit="false" (or remove it) to require an exact answer for any credit.
+        #  Set partial_credit="EDC" to count each choice for equal points (Every Decision Counts).
+        #  Set partial_credit="halves" to take half credit off for each error.
+
+        tree = self.xml
+        problem_xml = tree.xpath('.')
+
+        # Partial credit type - only one type at a time right now.
+        credit_type = problem_xml[0].get('partial_credit', default=False)
+
+        try:
+            credit_type = str(credit_type).lower().strip()
+        except ValueError:
+            _ = self.capa_system.i18n.ugettext
+            # Translators: 'partial_credit' is an attribute name and should not be translated.
+            # 'EDC' and 'halves' and 'false' should also not be translated.
+            msg = _("partial_credit value should be one of 'EDC', 'halves', or 'false'.")
+            raise LoncapaProblemError(msg)
+
+        if credit_type == 'false':
+            pass
+        elif credit_type == 'halves':
+            return self.grade_via_halves(all_choices, student_answer, student_non_answers)
+        elif credit_type == 'edc':
+            return self.grade_via_edc(all_choices, student_answer, student_non_answers)
+
```

Comment:
I'm not sure your background in Python. That's a fairly standard Pythonic way to do this, and makes for much shorter, cleaner code. However, it might be a little opaque if you're coming in from a different background (esp. if from non-functional languages). 

As an offer (your call as to whether this would be helpful): I'm at AI-ED+EDM right now, but if you wanted a quick tutorial on the relevant parts of Python, or to hack on this together, I'd be glad to volunteer my time. I'm a little booked up when I first get back, but I'll be available from 3pm Thursday next week, and fairly clear for most of the rest of July.